Automatically detect Chrome browser binary if it exists

DanielRyanSmith opened the pull request:
The current implementation of the Chrome class does not check the specified virtualenv path for an existing Chrome browser binary. Passing in the `venv_path` when trying to find the binary will ensure the binary is found properly if it exists.

~~Also, currently a generic error "Unable to locate Chrome binary" will be thrown when running tests against Chrome via `wpt run chrome` command with no browser binary in the virtualenv directory. Because a browser binary must exist in the virtualenv directory and not just the local machine, a more verbose error message has been added to explain how to fix the issue.~~
**EDIT:** This part is more of a separate issue, and can be fixed at a later time.

DanielRyanSmith committed:
automatically detect chrome browser binary if exists

jgraham commented on tools/wpt/browser.py:
```diff
@@ -633,7 +633,9 @@ def find_nightly_binary(self, dest):
         if channel == "nightly":
-            return self.find_nightly_binary(self._get_dest(venv_path, channel))
+            nightly_binary = self.find_nightly_binary(self._get_dest(venv_path, channel))
+            if nightly_binary is not None:
```
So the upshot of this is that if we try to find a nightly binary and fail, we'll fall back to using just `google-chrome` on the `$PATH` on linux and similar default names on other channels? It's unclear to me if that's expected.

DanielRyanSmith replied:
That is the intent, the current implementation will just throw a generic error if a nightly binary doesn't exist, no falling back. But maybe change is not a good idea from some usability perspective and falling back could be an issue?

DanielRyanSmith replied:
It looks like Firefox has a similar implementation where the virtual environment is checked, and then falls back if a binary is not located there:

https://github.com/web-platform-tests/wpt/blob/ab2592d3c62f8f131d695e7c5c0ab08e273c85ce/tools/wpt/browser.py#L268-L285

jgraham replied:
The difference is that Firefox doesn't have different binary names for nightly vs stable. Since Chrome does, I'm not sure if it makes sense to use something that's known to not be a nightly build when the channel was specified as nightly.

DanielRyanSmith replied:
I see, thank you for the clarification. I am bringing this up because this was a pain point when I first tried to run the tests in this repository. Do you think it would be better to simply add a more verbose error message for this section? "nightly" is also the default choice when you do not specify a channel and run the tests, so this issue comes up whenever a user specifies nothing.

https://github.com/web-platform-tests/wpt/blob/ae28863b68291fa71320ccc577aa874fbb7009b7/tools/wpt/install.py#L27-L37

DanielRyanSmith replied:
I updated the code to no longer fall back to looking on the `$PATH` or other channels if a browser binary is not found on the virtualenv path. This should behave exactly as the current implementation except a more verbose error message will display on how to fix the issue.

DanielRyanSmith replied:
@jgraham The more I thought about this, the more it feels like a separate issue from the main change. The most important part of this PR is the `venv_path` change, which fixes an issue where the Chrome browser binary would go completely undetected even if it existed in virtualenv. I removed the above change, and if it's still important, I will return to it when I am certain of the correct approach.

DanielRyanSmith committed:
Chrome nightly browser binary locating fix

If a nightly browser binary is not found in the virtual environment, a more verbose error will be displayed instructing the user how to fix the issue.

DanielRyanSmith committed:
remove change for nightly browser binary detection

jgraham reviewed (approved):
Thanks!

Re the Chrome issue, I think we need to fix the channel handling overall. Currently I think we pick a default channel too early in the process when none is supplied, when really we should defer picking a channel until after we found a binary we're going to use, because often we can infer the channel from the binary. So for Chrome we'd handle the `channel=None` `binary=None` case by looking for all possible binaries and picking one to use, and then setting the channel from the selected binary.

DanielRyanSmith commented:
> we should defer picking a channel until after we found a binary we're going to use, because often we can infer the channel from the binary. So for Chrome we'd handle the `channel=None` `binary=None` case by looking for all possible binaries and picking one to use, and then setting the channel from the selected binary.

@jgraham Do you think this is worth raising as a separate issue?
